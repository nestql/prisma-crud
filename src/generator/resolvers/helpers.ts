import { OptionalKind, MethodDeclarationStructure, Writers } from "ts-morph";

import { DmmfDocument } from "../dmmf/dmmf-document";
import { DMMF } from "../dmmf/types";

export function generateCrudResolverClassMethodDeclaration(
  action: DMMF.Action,
  mapping: DMMF.ModelMapping,
  dmmfDocument: DmmfDocument,
): OptionalKind<MethodDeclarationStructure> {
  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${action.returnTSType}>`,
    decorators: [
      {
        name: `TypeGraphQL.${action.operation}`,
        arguments: [
          `_returns => ${action.typeGraphQLType}`,
          Writers.object({
            nullable: `${!action.method.isRequired}`,
          }),
        ],
      },
    ],
    parameters: [
      {
        name: "ctx",
        // TODO: import custom `ContextType`
        type: "any",
        decorators: [{ name: "TypeGraphQL.Context", arguments: [] }],
      },
      {
        name: "info",
        type: "GraphQLResolveInfo",
        decorators: [{ name: "TypeGraphQL.Info", arguments: [] }],
      },
      ...(!action.argsTypeName
        ? []
        : [
            {
              name: "args",
              type: action.argsTypeName,
              decorators: [{ name: "TypeGraphQL.Args", arguments: [] }],
            },
          ]),
    ],
    statements:
      action.kind === DMMF.ModelAction.aggregate
        ? [
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...transformFields(graphqlFields(info as any)),
            });`,
          ]
        : action.kind === DMMF.ModelAction.groupBy
        ? [
            /* ts */ ` const { _count, _avg, _sum, _min, _max } = transformFields(
              graphqlFields(info as any)
            );`,
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...Object.fromEntries(
                Object.entries({ _count, _avg, _sum, _min, _max }).filter(([_, v]) => v != null)
              ),
            });`,
          ]
        : [
            /* ts */ ` const { _count } = transformFields(
              graphqlFields(info as any)
            );
            return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...(_count && transformCountFieldIntoSelectRelationsCount(_count)),
            });`,
          ],
  };
}
