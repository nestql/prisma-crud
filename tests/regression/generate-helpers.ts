import { promises as fs } from "fs";
import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";
import createReadGeneratedFile, {
  ReadGeneratedFile,
} from "../helpers/read-file";

describe("helpers", () => {
  let outputDirPath: string;
  let readGeneratedFile: ReadGeneratedFile;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath("helpers-regression");
    await fs.mkdir(outputDirPath, { recursive: true });
    readGeneratedFile = createReadGeneratedFile(outputDirPath);
  });

  it("should properly emit helpers file", async () => {
    const schema = /* prisma */ `
      model Sample {
        id    Int     @id @default(autoincrement())
        text  String
      }
    `;

    await generateCodeFromSchema(schema, { outputDirPath });
    const helpersTSFile = await readGeneratedFile("/helpers.ts");

    expect(helpersTSFile).toMatchSnapshot("helpers");
  });

  describe("when contextPrismaKey generator option is set", () => {
    it("should properly generate `getPrismaFromContext`", async () => {
      const schema = /* prisma */ `
        model Sample {
          id    Int     @id @default(autoincrement())
          text  String
        }
      `;

      await generateCodeFromSchema(schema, {
        outputDirPath,
        contextPrismaKey: "samplePrisma",
      });
      const helpersTSFile = await readGeneratedFile("/helpers.ts");

      expect(helpersTSFile).toMatchSnapshot("helpers");
    });
  });
});
