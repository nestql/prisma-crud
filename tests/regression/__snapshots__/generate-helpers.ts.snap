// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers should properly emit helpers file: helpers 1`] = `
"
export function transformFields(fields: Record<string, any>): Record<string, any> {
  return Object.fromEntries(
    Object.entries(fields)
      // remove __typename and others
      .filter(([key, value]) => !key.startsWith(\\"__\\"))
      .map<[string, any]>(([key, value]) => {
        if (Object.keys(value).length === 0) {
          return [key, true];
        }
        return [key, transformFields(value)];
      }),
  );
}

export function getPrismaFromContext(context: any) {
  const prismaClient = context[\\"prisma\\"];
  if (!prismaClient) {
    throw new Error(\\"Unable to find Prisma Client in GraphQL context. Please provide it under the \`context[\\\\\\"prisma\\\\\\"]\` key.\\");
  }
  return prismaClient;
}

export function transformCountFieldIntoSelectRelationsCount(_count: object) {
  return {
    include: {
      _count: {
        select: {
          ...Object.fromEntries(
            Object.entries(_count).filter(([_, v]) => v != null)
          ),
        }
      },
    },
  }
}



"
`;

exports[`helpers when contextPrismaKey generator option is set should properly generate \`getPrismaFromContext\`: helpers 1`] = `
"
export function transformFields(fields: Record<string, any>): Record<string, any> {
  return Object.fromEntries(
    Object.entries(fields)
      // remove __typename and others
      .filter(([key, value]) => !key.startsWith(\\"__\\"))
      .map<[string, any]>(([key, value]) => {
        if (Object.keys(value).length === 0) {
          return [key, true];
        }
        return [key, transformFields(value)];
      }),
  );
}

export function getPrismaFromContext(context: any) {
  const prismaClient = context[\\"samplePrisma\\"];
  if (!prismaClient) {
    throw new Error(\\"Unable to find Prisma Client in GraphQL context. Please provide it under the \`context[\\\\\\"samplePrisma\\\\\\"]\` key.\\");
  }
  return prismaClient;
}

export function transformCountFieldIntoSelectRelationsCount(_count: object) {
  return {
    include: {
      _count: {
        select: {
          ...Object.fromEntries(
            Object.entries(_count).filter(([_, v]) => v != null)
          ),
        }
      },
    },
  }
}



"
`;
